File: frontend/src/lib/tzClock.ts
// The location's clock, computed once, by the composer (tide-weather-planner,
// schema section 5). Twin of backend/services/tz_clock.py; both are driven from
// the same fixture (tzClock.fixture.json) so a DST edge or a half-hour zone
// cannot be placed differently on the two transports.
//
// Every `ts` is an integer epoch SECOND and every arithmetic result is an
// integer; the wall-clock string is display-only and is never compared, sorted
// or bracketed by (schema section 5: one axis, real elapsed time). Pure: no
// clock is read here, `ts` is always an argument.
//
// Intl formatters are memoised per zone because the weather builder calls
// `localClock` for every day, event and cell of a plan, and constructing an
// Intl.DateTimeFormat is the expensive half of formatting.

const fmtCache = new Map<string, Intl.DateTimeFormat>()

function fmtFor(tz: string): Intl.DateTimeFormat {
  let f = fmtCache.get(tz)
  if (!f) {
    f = new Intl.DateTimeFormat('en-CA', {
      timeZone: tz,
      year: 'numeric', month: '2-digit', day: '2-digit',
      hour: '2-digit', minute: '2-digit', second: '2-digit',
      hourCycle: 'h23',
    })
    fmtCache.set(tz, f)
  }
  return f
}

interface WallParts { y: number; mo: number; d: number; h: number; mi: number; s: number }

function wallParts(ts: number, tz: string): WallParts {
  const parts = fmtFor(tz).formatToParts(new Date(ts * 1000))
  const g = (t: string) => Number(parts.find(p => p.type === t)?.value ?? '0')
  // Some ICU builds still spell midnight '24' under h23 for the hour part.
  const h = g('hour') === 24 ? 0 : g('hour')
  return { y: g('year'), mo: g('month'), d: g('day'), h, mi: g('minute'), s: g('second') }
}

const pad2 = (n: number) => String(n).padStart(2, '0')

/** The zone's UTC offset in seconds at instant `ts` (east positive). */
export function utcOffsetSec(ts: number, tz: string): number {
  const p = wallParts(ts, tz)
  return Math.round(Date.UTC(p.y, p.mo - 1, p.d, p.h, p.mi, p.s) / 1000) - ts
}

/** 'YYYY-MM-DD HH:MM' in the zone. Display only. */
export function localClock(ts: number, tz: string): string {
  const p = wallParts(ts, tz)
  return `${p.y}-${pad2(p.mo)}-${pad2(p.d)} ${pad2(p.h)}:${pad2(p.mi)}`
}

/** 'YYYY-MM-DD' in the zone: the first ten characters of `localClock`. */
export function localDate(ts: number, tz: string): string {
  return localClock(ts, tz).slice(0, 10)
}

/** The instant the zone's current hour began: `ts` rolled back to :00 on the
 *  LOCATION's clock, so a half-hour zone gets its own hour boundary. */
export function startOfLocalHour(ts: number, tz: string): number {
  const off = utcOffsetSec(ts, tz)
  return ts - ((((ts + off) % 3600) + 3600) % 3600)
}

/** 'YYYY-MM-DD' plus `days` calendar days, pure calendar arithmetic. */
export function addDays(date: string, days: number): string {
  const y = Number(date.slice(0, 4)), mo = Number(date.slice(5, 7)), d = Number(date.slice(8, 10))
  const t = new Date(Date.UTC(y, mo - 1, d + days))
  return `${t.getUTCFullYear()}-${pad2(t.getUTCMonth() + 1)}-${pad2(t.getUTCDate())}`
}

/**
 * The instant local midnight begins on `date` in the zone, as an integer epoch
 * second.
 *
 * A zone whose midnight is ambiguous (a fall-back at 01:00 makes 00:00 occur
 * twice) settles on the FIRST occurrence; a zone whose midnight does not exist
 * (a spring-forward at 00:00) settles on the first existing instant after the
 * gap. Both are exactly what Python's zoneinfo gives `fold=0`, which is what
 * the twin relies on, so the two composers place day boundaries identically.
 * No US zone changes its clock at midnight, so on every US fixture the three
 * candidates below agree and the answer is the one fixed point.
 */
export function localMidnightTs(date: string, tz: string): number {
  const y = Number(date.slice(0, 4)), mo = Number(date.slice(5, 7)), d = Number(date.slice(8, 10))
  const wall = Math.round(Date.UTC(y, mo - 1, d) / 1000)
  // The offsets in force half a day either side of the naive instant cover
  // every transition that could touch this midnight.
  const offsets = new Set<number>()
  for (const probe of [wall - 43200, wall, wall + 43200]) offsets.add(utcOffsetSec(probe, tz))
  const candidates = [...offsets].map(off => wall - off).sort((a, b) => a - b)
  const wanted = `${date} 00:00`
  const exact = candidates.filter(c => localClock(c, tz) === wanted)
  if (exact.length > 0) return exact[0]
  // A gap: no instant reads 00:00. fold=0 uses the offset in force BEFORE the
  // transition, which is the offset at the earliest candidate.
  return wall - utcOffsetSec(candidates[0], tz)
}

File: backend/services/tz_clock.py
"""The location's clock, computed once, by the composer (tide-weather-planner,
schema section 5). Twin of frontend/src/lib/tzClock.ts; both halves are driven
from the same fixture (frontend/src/lib/tzClock.fixture.json) so a DST edge or a
half-hour zone cannot be placed differently on the two transports.

Every `ts` is an integer epoch SECOND and every arithmetic result is an
integer; the wall-clock string is display-only and is never compared, sorted
or bracketed by. Pure: no clock is read here, `ts` is always an argument.

`gmt_epoch` places a NOAA `t` string requested in `time_zone=gmt` on the epoch
axis. It is now a thin unit conversion over the shared `place_instant` in
services/tide_instant.py rather than a second regex of its own.

THAT CONSOLIDATION CORRECTS A CLAIM THIS DOCSTRING CARRIED. It used to say
`gmt_epoch` was "deliberately not the shipped `_epoch_min` in services/tide.py,
whose naive `datetime.timestamp()` reads the PROCESS zone" -- true when written
and false since v1.0.32, which gave `_epoch_min` the explicit `tzinfo=utc` for
exactly that reason. From that point the two were the same UTC calendar
arithmetic behind two byte-identical patterns, differing only in their unit
(seconds here, minutes there) and in what they returned for a string they could
not read. Keeping them apart bought nothing and cost a divergence: `_epoch_min`
was fixed to refuse impossible calendar values where its TS twin rolled them
over, and the Planner pair inherited that same split. One predicate, two unit
conversions.
"""

from datetime import datetime, timedelta
from zoneinfo import ZoneInfo

from services.tide_instant import place_epoch_sec


def utc_offset_sec(ts: int, tz: ZoneInfo) -> int:
    """The zone's UTC offset in seconds at instant `ts` (east positive)."""
    return int(datetime.fromtimestamp(ts, tz).utcoffset().total_seconds())


def local_clock(ts: int, tz: ZoneInfo) -> str:
    """'YYYY-MM-DD HH:MM' in the zone. Display only."""
    return datetime.fromtimestamp(ts, tz).strftime("%Y-%m-%d %H:%M")


def local_date(ts: int, tz: ZoneInfo) -> str:
    """'YYYY-MM-DD' in the zone: the first ten characters of `local_clock`."""
    return local_clock(ts, tz)[:10]


def start_of_local_hour(ts: int, tz: ZoneInfo) -> int:
    """The instant the zone's current hour began, on the LOCATION's clock."""
    off = utc_offset_sec(ts, tz)
    return ts - (((ts + off) % 3600) + 3600) % 3600


def add_days(date: str, days: int) -> str:
    """'YYYY-MM-DD' plus `days` calendar days, pure calendar arithmetic."""
    y, m, d = (int(x) for x in date.split("-"))
    return (datetime(y, m, d) + timedelta(days=days)).strftime("%Y-%m-%d")


def local_midnight_ts(date: str, tz: ZoneInfo) -> int:
    """The instant local midnight begins on `date` in the zone.

    zoneinfo's fold=0 semantics: an ambiguous midnight resolves to its FIRST
    occurrence and a non-existent one to the first existing instant after the
    gap. The TS twin reproduces exactly this, so both composers place day
    boundaries identically."""
    y, m, d = (int(x) for x in date.split("-"))
    return int(datetime(y, m, d, tzinfo=tz).timestamp())


def gmt_epoch(t: str) -> int | None:
    """A NOAA 'YYYY-MM-DD HH:MM' string requested in `time_zone=gmt`, as an
    integer epoch second; None for a string that names no instant (the caller
    drops it).

    NONE RATHER THAN 0, because the caller's test was `if t > 0` and that is not
    the predicate it meant. `1970-01-01 00:00` is a placeable instant whose
    epoch IS the sentinel, and a pre-1970 instant is placeable and negative, so
    the shipped Planner silently dropped both. The predicate is "placeable", not
    "positive". Linearity, the character-class rule and the calendar round trip
    all live at `place_instant`, which this delegates to.
    """
    return place_epoch_sec(t)
